### ISSUE TITLE
Framebuffer compatibility with RenderPasses

### ISSUE THREAD

**Issue by schwa423** [KhronosGroup/Vulkan-Docs#404]
This originally started as issue #403, but the discussion there made it clear that there are two issues:
- a lack of clarity in section 7.2 "Render Pass Compatibility"
- misleading or insufficient language in section 7.3 "Framebuffers"

The first topic is covered by issue #403.  This new issue is for the second topic.  Consider the following scenario:

RenderPass1 is created with:
- pAttachments = [depth, color]
- Subpass 0: pColorAttachments[0].attachment = 1, pDepthStencilAttachment[0].attachment = 0

RenderPass2 is created with:
- pAttachments = [depth, color1, color2]
- Subpass 0: pColorAttachments[0].attachment = 2, pDepthStencilAttachment[1].attachment = 0

Framebuffer is created with:
- renderPass = RenderPass1
- pAttachments = [depthImageView, colorImageView]

According to section 7.2 "Render Pass Compatibility", RenderPass1 and RenderPass2 are compatible.

According to section 7.3 "Framebuffers", `VkFramebufferCreateInfo.renderPass` is described as:
> a render pass that defines what render passes the framebuffer will be compatible with. See Render Pass Compatibility for details.

This seems to suggest that Framebuffer can be used as the target of RenderPass2.

However, the validation function `VerifyFramebufferAndRenderPassLayouts()` (see https://gitlab.khronos.org/vulkan/LoaderAndValidationLayers/blob/1.0/layers/core_validation.cpp#L10391) seems to disagree with this, in particular the code:

```C++
if (pRenderPassInfo->attachmentCount != framebufferInfo.attachmentCount) {
        skip_call |= log_msg(dev_data->report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, (VkDebugReportObjectTypeEXT)0, 0, __LINE__,
                             DRAWSTATE_INVALID_RENDERPASS, "DS", "You cannot start a render pass using a framebuffer "
                                                                 "with a different number of attachments.");
}
```

If RenderPass1 and RenderPass2 are, in fact, compatible according to section 7.2 "Render Pass Compatibility", then section 7.3 "Framebuffers" is misleading at best.

**Comment by krOoze** [comment]
Sorry to burst your bubble, but `VkFramebufferCreateInfo` VU again:

> -  `attachmentCount` **must** be equal to the attachment count specified in `renderPass` 

Your example would not be a Valid Vulkan Application.

That's how it goes in specifications. It is not sufficient to complay to **any** _requirement_. It is mandatory to comply to **all** requirements at all times.
If it violates one requirement, then "compatibility" or "no compatibility" is besides the point.

**Comment by schwa423** [comment]
When reading section 7.4 "Render Pass Commands", it becomes clear that some (all?) variants of the scenario above are not valid usage.  

For example, if the initial/final layout of the color attachment in RenderPass2 is `VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL`, then the "corresponding attachment image subresource" (meaning, I presume, the one at the same index) must have been created with `VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT` set.

However, it is not immediately evident whether there is some combination of settings that aren't (accidentally?) caught by the conditions for valid usage of `VkRenderPassBeginInfo`.  

If scenarios such as the one above are intended to be illegal, it would be better if this was explicit in the wording of sections 7.2 and/or 7.3, regardless of whether the union of valid usage conditions in other sections rules it out (especially when the current wording of section 7.3 directs the reader to 7.2 for further details, and omits 7.4).

**Comment by schwa423** [comment]
@krOoze:

Burst away!  However, I think you're wrong in this case.  The scenario I describe is a valid usage of `VkFramebufferCreateInfo`, because Framebuffer is created with RenderPass1 and a list of attachments that match those described by RenderPass1.

**Comment by krOoze** [comment]
@schwa423 Oh, I think you are right!

One would expect the same requirements for `VkFramebufferCreateInfo` to be raised on the Compatible substitute( which does not have to go through `VkFramebufferCreateInfo` ), right?

**Comment by schwa423** [comment]
@krOoze 

Yes, correct.  This is why I was wondering in #403 whether corresponding Attachment References should "have the same index"... by tightening up the requirements for "Render Pass Compatibility", the desired validity constraints might be enforced based solely on the wording of section 7.2, without additional clarifying text in section 7.3 and 7.4.

**Comment by krOoze** [comment]
Depending how #403 is resolved, there may be some problem with Image Usage requirements too (as raised in `VkFramebufferCreateInfo` VU).

If any rewiring through References is allowed, then that is broken (otherwise if not then the Render Pass Compatibility sure goes round-about way to explain that)

**Comment by janharaldfredriksen-arm** [comment]
I think this is covered by the fix for #403.

**Comment by CarloWood** [comment]
@janharaldfredriksen-arm Can you point me to where this is addressed exactly in the current spec? Because I don't see it - seems that the exact same problem still exists.
